### ISSUE TITLE
HipChat requires Oauth2 now

### ISSUE THREAD

**Issue by garlandkr** [gitlabhq/gitlabhq#7373]
When adding a token to a room and running the test message there is a resulting error:

```
Started GET "test/services/hipchat/test" for x.x.x.x at 2014-07-25 19:42:04 +0000
Processing by Projects::ServicesController#test as HTML
  Parameters: {"project_id"=>"test", "id"=>"hipchat"}
Completed 500 Internal Server Error in 135ms

HipChat::Unauthorized (Access denied to room `test'):
  app/models/project_services/hipchat_service.rb:45:in `execute'
  app/controllers/projects/services_controller.rb:29:in `test'


Started PUT "test" for x.x.x.x at 2014-07-25 19:42:10 +0000
Processing by Projects::ServicesController#update as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"test", "service"=>{"active"=>"1", "token"=>"test", "room"=>"test"}, "project_id"=>"test", "id"=>"hipchat"}
Redirected to https://test/services/hipchat/edit
Completed 302 Found in 40ms (ActiveRecord: 6.5ms)
Started GET "test/services/hipchat/edit" for x.x.x.x at 2014-07-25 19:42:10 +0000
Processing by Projects::ServicesController#edit as HTML
  Parameters: {"project_id"=>"test", "id"=>"hipchat"}
Completed 200 OK in 83ms (Views: 45.8ms | ActiveRecord: 5.9ms)
Started GET "test/services/hipchat/test" for x.x.x.x at 2014-07-25 19:42:11 +0000
Processing by Projects::ServicesController#test as HTML
  Parameters: {"project_id"=>"test", "id"=>"hipchat"}
Completed 500 Internal Server Error in 135ms

HipChat::Unauthorized (Access denied to room `test'):
  app/models/project_services/hipchat_service.rb:45:in `execute'
  app/controllers/projects/services_controller.rb:29:in `test'
```

Account credentials need to be added somewhere so an authenticated request can be made from the GitLab settings page for the repo.

**Comment by jerrac** [comment]
Just ran into this issue here at work. Is there a way to tell how many people have subscribed to this issue? If we just look at how many people have commented, there aren't many who are using the HipChat integration...

**Comment by aangov** [comment]
Confirming the same issue. Gitlab is not usable with HipChat anymore.

**Comment by jamek** [comment]
https://[yoursubdomain].hipchat.com/admin/api here you generate token which works with gitlab  as it works in my company.

**Comment by aangov** [comment]
Tested it with several projects, both with commits and by using the built-in "Test" button. Getting the same in the production.log for all of them:

Started GET "/project/bla/bla/hipchat/test" for 31.11.94.125 at 2015-01-13 12:32:56 +0000
Processing by Projects::ServicesController#test as HTML
  Parameters: {"project_id"=>"project/id", "id"=>"hipchat"}
Completed 500 Internal Server Error in 61ms

HipChat::Unauthorized (Access denied to room `test'):
  app/models/project_services/hipchat_service.rb:43:in `execute'
  app/controllers/projects/services_controller.rb:29:in `test'

**Comment by boomy** [comment]
Confirmed, get the same error.

**Comment by siebertm** [comment]
Today I upgraded to 7.7 (omnibus) and the hipchat integration stopped working.

What needs to be done to make it working again?

**Comment by irobayna** [comment]
same issue on 7.7.0

**Comment by chulkilee** [comment]
You can keep using existing group-level API key with v1 api. However api version is hardcoded in gitlab, so #8645 will fix it.

**Comment by frenck** [comment]
Simple (but very nasty) workaround for us is to force the HipChat service back to the v1 api.

https://github.com/gitlabhq/gitlabhq/blob/master/app/models/project_services/hipchat_service.rb#L49

Change v2 to v1 and restart (gitlab-ctl restart).

In omnibus installations this file is found in:
/opt/gitlab/embedded/service/gitlab-rails/app/models/project_services

**Comment by siebertm** [comment]
@frenck: ugly but works like a charm.

**Comment by Soullivaneuh** [comment]
@frenck Thanks. You should make a PR of that, no?

**Comment by stanhu** [comment]
I was a little puzzled why people were having trouble, since I've been doing a lot of work on improving the HipChat notifications and having no issues.  Here is the lowdown:

HipChat v1 API (legacy) supports "API Auth Tokens" in the Group API menu. A v1 token is allowed to send messages to any room.

HipChat v2 API has tokens that are can be created using the *Integrations* tab in the Group or Room admin page. You can create a custom integration that has a URL of something like:

```
https://api.hipchat.com/v2/room/<room>/notification?auth_token=<token>
```

OAuth2 is unnecessary, since the HipChat API is designed to work with a pre-existing token as before.

Atlassian has already deprecated v1, and I think specifying the version number in the Services panel  would just lead to confusion. Please correct me if I'm wrong, but I think it's better to mandate the use of the v2 API and tokens.

**Comment by siebertm** [comment]
@stanhu then maybe we're trying to set it up the wrong way... Is there any guide we could follow? Maybe link or integrate that in the settings?

**Comment by irobayna** [comment]
@stanhu I would agree if v2 were to work, it does not as is --at least not for me.   Following @frenck  recommendation works

**Comment by stanhu** [comment]
@siebertm @irobayna Do these instructions work?

1) Go to: https://admin.hipchat.com/admin
2) Click on "Group Admin" -> "Integrations".
3) Find "Build Your Own!" and click "Create".
4) Select the desired room, name the integration "GitLab", and click "Create".
5) In the "Send messages to this room by posting this URL" column, you should see a URL in the format:
```
https://api.hipchat.com/v2/room/<room>/notification?auth_token=<token>
```
6) Take the `token` field and put it into the `Token` field in the GitLab HipChat service.

**Comment by siebertm** [comment]
7) Take the `room` field and put it in the `Room` field in the Gitlab HipChat service.

Works like a charm! At least for our installation I can say: I configured it the wrong way.

Thanks @stanhu ! IMO the issue should be closed.

**Comment by stanhu** [comment]
@siebertm great to hear! FYI, in GitLab 7.9, which will be released on March 22nd, the Room field has been replaced by the Channel field, since you can also send private messages to a user. It shouldn't be necessary to specify a room unless you want to override the integration default. That's why I didn't mention it.

**Comment by chulkilee** [comment]
@stanhu Are you sure v2 API token works with **any** rooms? From my testing I'm getting "Access denied to room X" when I tried to use for different room other than the one I specified when created it.

I may create keys for each room and update existing service configuration accordingly, but I think it is not ideal. It also breaks existing configuration so you cannot simply mandate users to use v2 token. gitlab admins using hipchat should know what v1 and v2 key are (and their difference) and be able to pick one that work for their use cases.

I agree that eventually hipchat service integration should drop support for deprecated API version (v1), though. Actually I expect better OAuth-based integration between gitlab and hipchat, but I'm not sure it's easy in the current project service model.

Until then I'll use #8645 to keep using v1 key.

**Comment by stanhu** [comment]
@chulkilee By design, the room integration token is supposed to be  a lightweight token for one specific room (see https://www.hipchat.com/docs/apiv2/auth). 

If you want to be able to specify arbitrary rooms, you can create an API token for a specific user under "Account settings" and "API access". Use the `XXX` value under `auth_token=XXX`.

**Comment by irobayna** [comment]
@stanhu works!! :+1: 
This is great! I am guessing the problem was a configuration issue on my part. Thank you for the instructions.  Please lets make them available somewhere :-)

**Comment by chulkilee** [comment]
Okay so there are three types of API keys actually.

- V1 group-level API key (from API tab)
- V2 group-level API key (from Integration tab) tied to a room
- V2 account-level API key (from User Settings -> API access)

They work differently and belong to different entities, so I'm still against forcing users to use V2 :(

**Comment by stanhu** [comment]
Closed via 57c724558db.
